### ISSUE TITLE
Version handle the API

### ISSUE THREAD

**Issue by m-linner-ericsson** [eiffel-community/eiffel-intelligence#427]
### Description
Introduce version handling of the API

### Motivation
By version handling, the API users can upgrade EI without having to change their code. The user can switch to the new API when they have the possibility.

### Benefits
By version handling the API, it will be backward compatible.

**Comment by e-pettersson-ericsson** [comment]
Is the current API (from 3.0.0 release) going to be the default and v1 of the API? 
And any changes made from here on out should be backwards compatible, and if they're not they should be a v2?

I'd like this to be a backwards-compatible change to 3.0.0 release. (Since we've already made breaking changes between 2.x and 3.x release) Meaning that this current API will be the default if users don't specify any version in the uri.

`http://eiffel-intelligence/subscriptions` and `http://eiffel-intelligence/v1/subscriptions` should point to the same default version of the API.

In the future, if users want another version than default they include it in the uri, e.g. http://eiffel-intelligence/v2/subscriptions etc

**Comment by e-pettersson-ericsson** [comment]
I noticed the possibility of grouping different API's with Springfox implementation of Swagger. I don't know if this is somehting we want, but whoever looks into this issue can perhaps take a look and see if it's useful. Read more here: http://springfox.github.io/springfox/docs/2.6.1/#swagger-group

**Comment by tobiasake** [comment]
Sounds as good solution to versioning the API.
I have seen issues that users use entrypoints that is available on root url and when a new version of application with rest entrypoint changes, then users is impacted and the users application functionality stop working.

My suggestion:
1. Don't have any entrypoints on the root url. User should specify a specific API version from the start, so the users application is not impacted when a new version is deployed with new API changes(that introduces a new API version).
2. EI 2.X RestApi entrypoints should be placed under "v1/" root url context-path.
3. EI 3.X RestApi entrypoints should be placed under "v2/" root url context-path.
4. EI application root-url without API version should redirect to Swagger page.

By implementing according these 4 implementation suggestions, we will avoid that users will be impacted by RestApi changes when new EI versions is deployed and users applications can decide when its suitable to switch to new API version.

**Comment by tobiasake** [comment]
How should we deprecate entrypoints? How long or many releases should we keep deprecated entrypoints versions?

We could have a rule to only keep 2 deprecated API versions:
Example:
v4/
v3/ (deprecated)
v2/ (deprecated)
v1/ (deprecated) -> removed when version /v4 is released.

Next version:
v5/
v4/ (deprecated)
v3/ (deprecated)
v2/ (deprecated) -> removed when version /v5 is released.

What do you think? or only one deprecated version be kept?

**Comment by zrddnvt** [comment]
Yeah, if I understood correctly, If EI 2.X RestApi entrypoints should be placed under "v1/" root url context-path and EI 3.X RestApi entrypoints should be placed under "v2/" root url context-path. then every entrypoint and its functionality have to be added to every future versions without any change. So instead we have approached through @Mapping()
EX: @GetMapping(value = {"/v1","/v2"}).
Swagger grouping is done to the respective entrypoints or its required functionality. 
So that mapping could be done to its respective entrypoints if there is a change in version.

**Comment by tobiasake** [comment]
Yes, user application need to have version specified to keep the backward compatibility and avoid user to be impacted by API changes.

**Comment by zrddnvt** [comment]
overview: 

**According to the version handling restAPI Endpoints :**

Version handling is to continue using the existing REST API and migrate their applications to the newer API when they are ready.
-> Identified restAPI endpoints between the versions.
-> Identified changes between the versions and covered the user scenarios as follows:

- Add new endpoints
- Delete endpoints
- Endpoint name change
- Endpoint request params(Header, Path, Query string Request body parameters) change.
- Endpoint functionality change
- Endpoint response changes. 
so based on the change between the versions by swagger group concept and map the versions using @mapping(). Ex: @GetMapping(value = {"/v1","/v2"}) to its respective changed functionality of the version, where old version kept in parallel.

Considered latest as the default version ( no URL change for the user ).
EX: If /v1  has changes in the endpoint and changes have been released in /v2 version then v2 will be considered as the default version because its the latest one.
User need the change the urls if requires old versions entrypoints.

**Comment by tobiasake** [comment]
Latest API version link to api version "v2/" has very bad impact on the users, since user will get API version "v3/" when new EI version is deployed, and things stops working for the user.
Due to this issue we should not have any entrypoints on root url and User should pick a API version "v1/" or "/v2" to use, so when API version "v3/" is deployed, user will not be impacted. User can continue use previous API versions as user did with previous EI versions.

**Comment by zrddnvt** [comment]
Yes as per our proposal if user picks a API version "v1/" or "v2/"  to use so when the new API "v3/" is deployed user will not be impacted. User can continue use previous API versions as user did with previous EI versions. User can then adapt or choose to the new API version when user find a suitable time for that. As in parallel we are maintaining previous versions.

**Comment by tobiasake** [comment]
WHat do you mean with previous versions? EI versions or the new API version structures versions?

**Comment by e-pettersson-ericsson** [comment]
So I interpret this as: we will not support a use case where the user accesses the API without any specified version (no default should exist). Users will be forced to always explicitly specify which version of the API they want to access - is this correct?

Example:

Allowed:
www.eiffel-intelligence/v1/subscriptions
www.eiffel-intelligence/v2/subscriptions

Not allowed:
www.eiffel-intelligence/subscriptions

 I prefer it this way, but I just want to make sure we are in agreement on this. @zrddnvt @tobiasake

**Comment by tobiasake** [comment]
Yes,, thats my opinion and what also has been discussed in other applications developments I am part of.

**Comment by e-pettersson-ericsson** [comment]
Just FYI, when we implement this, we will break backwards compatibility yet again, since we will force existing users to start using an explicit version when they use EI REST API. We will have to make this change part of a 4.0 release of Eiffel Intelligence which will support 2 API versions (old one in 2.x and new in 3.x)

**Comment by tobiasake** [comment]
That we will do whatever we do,, so lets do it right from the beginning.

**Comment by e-pettersson-ericsson** [comment]
Should we also add the "deprecated" flag onto the old API with this implementation? We will still support the old API for some time. But to make users aware that they should migrate to the new API. And the old API will be kept until a new MAJOR release of Eiffel Intelligence (i.e. 5.x)?

**Comment by zrddnvt** [comment]
> WHat do you mean with previous versions? EI versions or the new API version structures versions?

EI versions.

**Comment by zrddnvt** [comment]
> So I interpret this as: we will not support a use case where the user accesses the API without any specified version (no default should exist). Users will be forced to always explicitly specify which version of the API they want to access - is this correct?
> 
> Example:
> 
> Allowed:
> [www.eiffel-intelligence/v1/subscriptions](http://www.eiffel-intelligence/v1/subscriptions)
> [www.eiffel-intelligence/v2/subscriptions](http://www.eiffel-intelligence/v2/subscriptions)
> 
> Not allowed:
> [www.eiffel-intelligence/subscriptions](http://www.eiffel-intelligence/subscriptions)
> 
> I prefer it this way, but I just want to make sure we are in agreement on this. @zrddnvt @tobiasake

okay, we actually considered latest as default so that, if user is not specifying any version explicitly.  
Example: 
www.eiffel-intelligence/subscriptions/v1 --- EI2.x version rest API
www.eiffel-intelligence/subscriptions----- EI3.x version rest API

If this is the approach required we will do that.
www.eiffel-intelligence/v1/subscriptions
www.eiffel-intelligence/v2/subscriptions

**Comment by tobiasake** [comment]
I don't think we should keep root url version,, because then we get another release with more new users that have same issue, where API is changed. My suggestion is to remove the root url entrypoint directly let user from start/early as possible use the versioned entrypoints.

Regardinh deprecated, I added a suggestion regarding that earlier in this thread.

**Comment by e-pettersson-ericsson** [comment]
> > So I interpret this as: we will not support a use case where the user accesses the API without any specified version (no default should exist). Users will be forced to always explicitly specify which version of the API they want to access - is this correct?
> > Example:
> > Allowed:
> > [www.eiffel-intelligence/v1/subscriptions](http://www.eiffel-intelligence/v1/subscriptions)
> > [www.eiffel-intelligence/v2/subscriptions](http://www.eiffel-intelligence/v2/subscriptions)
> > Not allowed:
> > [www.eiffel-intelligence/subscriptions](http://www.eiffel-intelligence/subscriptions)
> > I prefer it this way, but I just want to make sure we are in agreement on this. @zrddnvt @tobiasake
> 
> okay, we actually considered latest as default so that, if user is not specifying any version explicitly.
> Example:
> [www.eiffel-intelligence/subscriptions/v1](http://www.eiffel-intelligence/subscriptions/v1) --- EI2.x version rest API
> [www.eiffel-intelligence/subscriptions-----](http://www.eiffel-intelligence/subscriptions-----) EI3.x version rest API
> 
> If this is the approach required we will do that.
> [www.eiffel-intelligence/v1/subscriptions](http://www.eiffel-intelligence/v1/subscriptions)
> [www.eiffel-intelligence/v2/subscriptions](http://www.eiffel-intelligence/v2/subscriptions)

Exactly like @tobiasake mentioned above. Let's imagine if we have the following alternatives to the API:

www.eiffel-intelligence/v1/subscriptions (old API)
www.eiffel-intelligence/v2/subscriptions (latest API)
www.eiffel-intelligence/subscriptions (default linking to latest version)

When we add a new API version: `www.eiffel-intelligence/v3/subscriptions` 
this will be the latest, and users currently using `www.eiffel-intelligence/subscriptions` 
without any version specified will notice things start breaking when they upgrade to new version of EI and wonder why. This is because the default API (without explicit version) will point to a NEW latest. 

This pattern will continue happening if we allow users to access the API without a version.

**Comment by e-pettersson-ericsson** [comment]
> How should we deprecate entrypoints? How long or many releases should we keep deprecated entrypoints versions?
> 
> We could have a rule to only keep 2 deprecated API versions:
> Example:
> v4/
> v3/ (deprecated)
> v2/ (deprecated)
> v1/ (deprecated) -> removed when version /v4 is released.
> 
> Next version:
> v5/
> v4/ (deprecated)
> v3/ (deprecated)
> v2/ (deprecated) -> removed when version /v5 is released.
> 
> What do you think? or should only one deprecated version be kept?

If we wil lsupport 2 deprecated versions of the API this means we will support in total 3 API versions. How much more complexity does this give than only supporting 2 versions? I'm not sure what I think actually.

**Comment by tobiasake** [comment]
It was my first suggestion, I have not strong opinion about that.
We can have only 2 API version
- Current API version v2/
- Deprecated API version v1/

When EI with API version v3/ released:
- Current API version v3/
- Deprecated API version v2/
- Deprecated API version v1/ -> removed

Is that a better suggestion?

**Comment by e-pettersson-ericsson** [comment]
> It was my first suggestion, I have not strong opinion about that.
> We can have only 2 API version
> 
> * Current API version v2/
> * Deprecated API version v1/
> 
> When EI with API version v3/ released:
> 
> * Current API version v3/
> * Deprecated API version v2/
> * Deprecated API version v1/ -> removed
> 
> Is that a better suggestion?

I'd like to hear others opinion about this @zrddnvt @raja-maragani

**Comment by zrddnvt** [comment]
> It was my first suggestion, I have not strong opinion about that.
> We can have only 2 API version
> 
> * Current API version v2/
> * Deprecated API version v1/
> 
> When EI with API version v3/ released:
> 
> * Current API version v3/
> * Deprecated API version v2/
> * Deprecated API version v1/ -> removed
> 
> Is that a better suggestion?

Yes, as per my opinion this could be better. (Y)

**Comment by e-pettersson-ericsson** [comment]
I updated the description of the issue with what we've discussed. Does it seem OK?
